File: src/java/cdk/src/main/java/com/cdk/product/apiworker/ProductApiWorkerStack.java
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2024 Datadog, Inc.
 */

package com.cdk.product.apiworker;

import com.cdk.constructs.SharedProps;
import software.amazon.awscdk.Stack;
import software.amazon.awscdk.StackProps;
import software.amazon.awscdk.services.dynamodb.ITable;
import software.amazon.awscdk.services.dynamodb.Table;
import software.amazon.awscdk.services.secretsmanager.ISecret;
import software.amazon.awscdk.services.secretsmanager.Secret;
import software.amazon.awscdk.services.sns.ITopic;
import software.amazon.awscdk.services.sns.Topic;
import software.amazon.awscdk.services.ssm.StringParameter;
import software.constructs.Construct;

public class ProductApiWorkerStack extends Stack {
    public ProductApiWorkerStack(final Construct scope, final String id) {
        this(scope, id, null);
    }

    public ProductApiWorkerStack(final Construct scope, final String id, final StackProps props) {
        super(scope, id, props);

        ISecret ddApiKeySecret = Secret.fromSecretCompleteArn(this, "DDApiKeySecret", System.getenv("DD_SECRET_ARN"));
        
        String serviceName = "JavaProductApi";
        String env = System.getenv("ENV") == null ? "dev" : System.getenv("ENV");
        String version = System.getenv("VERSION") == null ? "dev" : System.getenv("VERSION");
        
        SharedProps sharedProps = new SharedProps(serviceName, env, version, ddApiKeySecret);

        String priceCalculatedTopicArn = StringParameter.valueForStringParameter(this, "/java/product-pricing/product-calculated-topic");
        ITopic priceCalculatedTopic = Topic.fromTopicArn(this, "PriceCalculatedTopic", priceCalculatedTopicArn);
        
        String tableName = StringParameter.valueForStringParameter(this, "/java/product-api/table-name");
        ITable table = Table.fromTableName(this, "ProductApiTable", tableName);
        
        var pricingService = new ApiWorkerService(this, "JavaProductApiWorker", new ApiWorkerServiceProps(sharedProps, priceCalculatedTopic, table));
    }
}

File: src/nodejs/lib/product-api-workers/product-api-worker-stack.ts
//
// Unless explicitly stated otherwise all files in this repository are licensed
// under the Apache License Version 2.0.
// This product includes software developed at Datadog (https://www.datadoghq.com/).
// Copyright 2024 Datadog, Inc.
//

import * as cdk from "aws-cdk-lib";
import { Secret } from "aws-cdk-lib/aws-secretsmanager";
import { Construct } from "constructs";
import { Datadog } from "datadog-cdk-constructs-v2";
import { SharedProps } from "../constructs/sharedFunctionProps";
import { ApiWorker } from "./api";
import { StringParameter } from "aws-cdk-lib/aws-ssm";
import { ITopic, Topic } from "aws-cdk-lib/aws-sns";

// no-dd-sa:typescript-best-practices/no-unnecessary-class
export class ProductApiWorkerStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const ddApiKey = Secret.fromSecretCompleteArn(
      this,
      "DDApiKeySecret",
      process.env.DD_SECRET_ARN!
    );

    const service = "NodeProductApi";
    const env = process.env.ENV ?? "dev";
    const version = process.env["COMMIT_HASH"] ?? "latest";

    const datadogConfiguration = new Datadog(this, "Datadog", {
      nodeLayerVersion: 115,
      extensionLayerVersion: 65,
      site: process.env.DD_SITE ?? "datadoghq.com",
      apiKeySecret: ddApiKey,
      service,
      version,
      env,
      enableColdStartTracing: true,
      enableDatadogTracing: true,
      captureLambdaPayload: true,
    });

    const sharedProps: SharedProps = {
      environment: env,
      serviceName: service,
      version,
      datadogConfiguration,
    };

    let priceCalculatedTopic: ITopic | undefined = undefined;

    const priceCalculatedTopicArn = StringParameter.fromStringParameterName(
      this,
      "PriceCalculatedTopicArn",
      "/node/product/pricing-calculated-topic"
    );
    
    priceCalculatedTopic = Topic.fromTopicArn(
      this,
      "PriceCalculatedTopic",
      priceCalculatedTopicArn.stringValue
    );

    const api = new ApiWorker(this, "ProductApiWorker", {
      sharedProps,
      ddApiKeySecret: ddApiKey,
      priceCalculatedTopic,
    });
  }
}
